Batteries/Data/Nat/Digits.lean

```diff
@@ -0,0 +1,303 @@
+
+The most significant digit is guaranteed to be nonzero. When the input number is zero, an empty
+list is returned.
+
+```
+toDigitsLE 123 10 = ([3,2,1] : List (Fin 10))
+toDigitsLE 2 2 = ([0,1] : List (Fin 2))
+toDigitsLE 0 12345 = ([] : List (Fin 12345))
+```
+-/
+def toDigitsLE (n base : Nat) (hbase : 2 ≤ base := by omega) : List (Fin base) :=
+  if hn : n = 0 then [] else
+    have : NeZero base := ⟨by omega⟩
+    have _hint : n / base < n := by
+      apply Nat.div_lt_of_lt_mul
+      rw [Nat.lt_mul_iff_one_lt_left (Nat.zero_lt_of_ne_zero hn)]
+      exact hbase
+    Fin.ofNat base n :: toDigitsLE (n / base) base
+
+theorem toDigitsLE_zero {h : 2 ≤ base} : toDigitsLE 0 base = [] := by
+  rw [toDigitsLE, dif_pos rfl]
+
+theorem toDigitsLE_ne_zero {n : Nat} {h : 2 ≤ base} [NeZero n] :
+    have : NeZero base := ⟨by omega⟩ -- TODO: yuck!
+    toDigitsLE n base = Fin.ofNat base n :: toDigitsLE (n / base) base := by
+  intros; rw [toDigitsLE, dif_neg (NeZero.ne n)]
+
+/-! # Big-Endian (`BE`) Functions -/
+
+/--
+Construct a natural number from a big-endian list of digits in the given base, represented
+as elements of `Fin base`.
+
+For example:
+```
+ofDigitsBE (base := 10) [2,3,4] = 234 -- 4*10^0 + 3*10^1 + 2*10^2
+ofDigitsBE (base := 2) [0,1,0,0] = 4 -- 0*2^0 + 0*2^1 + 1*2^2 + 0*2^3
+ofDigitsBE (base := b) [] = 0 -- any base, including "base 0" and "base 1"
+```
+
+There are no valid inputs other than `[]` in "base 0". For "base 1", the only valid "digit" is 0
+and therefore:
+```
+ofDigitsBE (base := 1) l = 0
+```
+-/
+def ofDigitsBE : List (Fin base) → Nat :=
+  loop 0
+where
+  /-- Tail-recursive loop for `ofDigitsBE` -/
+  loop
+  | n, [] => n
+  | n, d :: l => loop (n * base + d) l
+
+private theorem ofDigitsBE.loop_nil :
+    loop (base := base) n [] = n := rfl
+
+private theorem ofDigitsBE.loop_cons {l : List (Fin base)} :
+    loop n (d :: l) = loop (n * base + d) l := rfl
+
+private theorem ofDigitsBE.loop_eq {l : List (Fin base)} :
+    loop n l = n * base ^ l.length + loop 0 l := by
+  induction l generalizing n with
+  | nil => simp [loop_nil]
+  | cons d l ih =>
+    simp only [loop_cons, List.length_cons, Nat.zero_mul, Nat.zero_add]
+    rw [ih (n := n * base + d), ih (n := d), Nat.add_mul, Nat.add_assoc, Nat.mul_assoc,
+      ←Nat.pow_add_one']
+
+@[simp]
+theorem ofDigitsBE_nil : ofDigitsBE (base := base) [] = 0 := rfl
+
+@[simp]
+theorem ofDigitsBE_cons {l : List (Fin base)} :
+    ofDigitsBE (d :: l) = d * base ^ l.length + ofDigitsBE l := by
+  simp [ofDigitsBE, ofDigitsBE.loop_cons]
+  exact ofDigitsBE.loop_eq ..
+
+theorem ofDigitsBE_append {l₁ l₂ : List (Fin base)} :
+    ofDigitsBE (l₁ ++ l₂) = ofDigitsBE l₁ * base ^ l₂.length + ofDigitsBE l₂ := by
+  induction l₁ with
+  | nil => simp
+  | cons d₁ l₁ ih =>
+    simp +arith [ih, Nat.add_mul, Nat.pow_add', Nat.mul_assoc]
+
+@[simp]
+theorem ofDigitsBE_reverse_eq_ofDigitsLE {l : List (Fin base)} :
+    ofDigitsBE l.reverse = ofDigitsLE l := by
```

Comment:
I claim that because of this there really is no point having both versions of the function.

Reply:
There is:

* The LE version has an obvious user-friendly recursion pattern but is highly inefficient.
* The BE version is not user-friendly but it is naturally tail-recursive.
* The BE version is used via `csimp` to make an efficient version of the LE version.

Reply:
The names used here don't conflict with Mathlib at all. The expected blend-in is to use `toDigitsLE` (once `csimp` implemented as the reverse of `toDigitsBE`) as a more efficient replacement for Mathlib's `digitsAux`, the whole thing should be invisible to Mathlib.